sdk/identity/Azure.Identity/src/WindowsNativeMethods.cs

```diff
@@ -0,0 +1,136 @@
+            DOMAIN_VISIBLE_PASSWORD = 4,
+            MAXIMUM = 5
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct CredentialData
+        {
+            public uint Flags;
+            public CRED_TYPE Type;
+            public string TargetName;
+            public string Comment;
+            public System.Runtime.InteropServices.ComTypes.FILETIME LastWritten;
+            public uint CredentialBlobSize;
+            public IntPtr CredentialBlob;
+            public CRED_PERSIST Persist;
+            public uint AttributeCount;
+            public IntPtr Attributes;
+            public string TargetAlias;
+            public string UserName;
+        }
+
+        public const int ERROR_NOT_FOUND = 1168;
+
+        public const uint FORMAT_MESSAGE_ALLOCATE_BUFFER = 0x00000100;
+        public const uint FORMAT_MESSAGE_IGNORE_INSERTS = 0x00000200;
+        public const uint FORMAT_MESSAGE_FROM_SYSTEM = 0x00001000;
+
+        public static IntPtr CredRead(string target, CRED_TYPE type)
+        {
+            ThrowIfFailed(Imports.CredRead(target, type, 0, out IntPtr userCredential));
+            return userCredential;
+        }
+
+        public static void CredWrite(IntPtr userCredential) => ThrowIfFailed(Imports.CredWrite(userCredential, 0));
+
+        public static void CredDelete(string target, CRED_TYPE type) => ThrowIfFailed(Imports.CredDelete(target, type, 0));
+
+        public static void CredFree(IntPtr userCredential)
+        {
+            if (userCredential != IntPtr.Zero)
+            {
+                Imports.CredFree(userCredential);
+            }
+        }
+
+        private static void ThrowIfFailed(bool isSucceeded)
+        {
+            if (isSucceeded)
+            {
+                return;
+            }
+
+            var error = Marshal.GetLastWin32Error();
+            if (error != ERROR_NOT_FOUND)
+            {
+                throw new InvalidOperationException();
+            }
+
+            throw new InvalidOperationException(MessageFromErrorCode(error));
+        }
+
+        private static string MessageFromErrorCode(int errorCode)
+        {
+            // Twy Win32 first
+            uint flags = FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS;
+            IntPtr messageBuffer = IntPtr.Zero;
+
+            var length = Imports.FormatMessage(flags, IntPtr.Zero, errorCode, 0, ref messageBuffer, 0, IntPtr.Zero);
+            if (length == 0)
+            {
+                // If failed, try to convert NTSTATUS to Win32 error
+                int code = Imports.RtlNtStatusToDosError(errorCode);
+                return new Win32Exception(code).Message;
+            }
+
+            string message = null;
+            if (messageBuffer != IntPtr.Zero)
+            {
+                message = Marshal.PtrToStringUni(messageBuffer);
+                Marshal.FreeHGlobal(messageBuffer);
```

Comment:
I think try/finally might be appropriate here for messageBuffer  cleanup.

Reply:
Agreed.

Reply:
Fixed